flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SlotSharingExecutionSlotAllocator.java

```diff
@@ -0,0 +1,268 @@
+		}
+
+		SlotExecutionVertexAssignment createAssignment(ExecutionVertexID executionVertexId) {
+			Preconditions.checkArgument(executionSlotSharingGroup.getExecutionVertexIds().contains(executionVertexId));
+			SlotRequestId logicalSlotRequestId = new SlotRequestId();
+			CompletableFuture<LogicalSlot> logicalSlotFuture = allocateLogicalSlot(logicalSlotRequestId, executionVertexId);
+			SlotExecutionVertexAssignment assignment = new SlotExecutionVertexAssignment(executionVertexId, logicalSlotFuture);
+			requestedLogicalSlots.put(
+				executionVertexId,
+				logicalSlotRequestId,
+				assignment);
+			return assignment;
+		}
+
+		private CompletableFuture<LogicalSlot> allocateLogicalSlot(
+				SlotRequestId logicalSlotRequestId,
+				ExecutionVertexID executionVertexId) {
+			String logMessageBase = getLogicalSlotString(logicalSlotRequestId, executionVertexId);
+			LOG.debug("Request a {}", logMessageBase);
+			CompletableFuture<LogicalSlot> logicalSlotFuture = slotContextFuture.thenApply(physicalSlot -> {
+				LOG.debug("Allocated {}", logMessageBase);
+				return new SingleLogicalSlot(
+					logicalSlotRequestId,
+					physicalSlot,
+					null,
+					Locality.UNKNOWN,
+					this,
+					slotWillBeOccupiedIndefinitely);
+			});
+			logicalSlotFuture.exceptionally(cause -> {
+				LOG.debug("Failed {}", logMessageBase);
+				cancelLogicalSlotRequest(logicalSlotRequestId);
+				return null;
+			});
+			return logicalSlotFuture;
+		}
+
+		void cancelLogicalSlotRequest(ExecutionVertexID executionVertexID) {
+			cancelLogicalSlotRequest(requestedLogicalSlots.getKeyBByKeyA(executionVertexID));
+		}
+
+		void cancelLogicalSlotRequest(SlotRequestId logicalSlotRequestId) {
+			SlotExecutionVertexAssignment assignment = requestedLogicalSlots.getValueByKeyB(logicalSlotRequestId);
+			if (assignment != null) {
+				LOG.debug("Cancel {}", getLogicalSlotString(logicalSlotRequestId, assignment.getExecutionVertexId()));
+				assignment.getLogicalSlotFuture().cancel(false);
+				requestedLogicalSlots.removeKeyB(logicalSlotRequestId);
+			} else {
+				LOG.debug("No SlotExecutionVertexAssignment for logical {} from physical %s", logicalSlotRequestId);
+			}
+			if (requestedLogicalSlots.values().isEmpty()) {
+				sharedSlots.remove(executionSlotSharingGroup);
+				slotProvider.cancelSlotRequest(
+					physicalSlotRequestId,
+					new FlinkException("Slot is being returned from SlotSharingExecutionSlotAllocator."));
+			}
+		}
+
+		@Override
+		public void returnLogicalSlot(LogicalSlot logicalSlot) {
+			cancelLogicalSlotRequest(logicalSlot.getSlotRequestId());
+		}
+
+		@Override
+		public void release(Throwable cause) {
+			for (SlotExecutionVertexAssignment assignment : requestedLogicalSlots.values()) {
+				ExecutionVertexID executionVertexId = assignment.getExecutionVertexId();
+				LOG.debug("Release {}", getLogicalSlotString(requestedLogicalSlots.getKeyBByKeyA(executionVertexId), executionVertexId));
+				assignment
```

Comment:
Maybe add some comments here to explain that all the logical slot futures should have already been completed normally when `release` is invoked? Or even add some sanity checks? It took me some time to reason this assumption.

Reply:
Do they have to complete normally?
The method does not look to assume this. It releases the logical slot only if its future is done:
`.thenAccept(logicalSlot -> logicalSlot.release(cause));`
I thought that `logicalSlot.release` will eventually trigger `returnLogicalSlot/cancelLogicalSlotRequest`

Reply:
When `SharedSlot#release(...)` is invoked, it indicates that the physical slot was successfully allocated, otherwise the shared slot is not assigned as the payload yet. This further means the logical slot futures have also completed normally. This is how I understand the assumption.

If we do not have the assumption the logical slot futures have completed normally, we need to cancel the `slotContextFuture` as well as all the pending logical slot futures in this `release()` method.

Reply:
ok, I also expected that the release call only comes from an allocated physical slot, so no need to touch it and logical slots. I was confused because the code in this method does not depend on this assumption.
I will add the sanity checks to make it explicit.